// @salla.sa/twilight-components/dist/collection/components/salla-installment/salla-installment.js
/*!
 * Crafted with ❤ by Salla
 */
import { Host, h } from '@stencil/core';
export class SallaInstallment {
  constructor() {
    this.tabbyBorderRemoved = false;
    this.tabbyRemoveBorderTries = 0;
    this.price = undefined;
    this.language = salla.config.get('user.language_code');
    this.currency = salla.config.get('user.currency_code');
    this.tamaraIsActive = undefined;
    this.tabbyIsActive = undefined;
    this.spotiiIsActive = undefined;
    salla.lang.onLoaded(() => {
      const installment = salla.config.get('store.settings.installments');
      if (installment) {
        this.tamaraIsActive = installment.includes('tamara_installment');
        this.tabbyIsActive = installment.includes('tabby_installment');
        this.spotiiIsActive = installment.includes('spotii_pay');
      }
      this.renderInstallments();
    });
  }
  render() {
    return (h(Host, null, this.tamaraIsActive ?
      h("div", { class: "tamara-product-widget", "data-price": this.price, "data-currency": this.currency, "data-lang": this.language, "data-payment-type": "installment" })
      : '', this.tabbyIsActive ?
      h("div", { id: "tabbyPromoWrapper" }, h("div", { id: "tabbyPromo" }))
      : '', this.spotiiIsActive ?
      h("div", { class: "spotii-wrapper" }, h("div", { class: "spotii-promo" }))
      : ''));
  }
  renderInstallments() {
    // Tamara
    if (this.tamaraIsActive) {
      var script = document.createElement('script');
      script.setAttribute('src', 'https://cdn.tamara.co/widget/product-widget.min.js');
      document.head.appendChild(script);
      script.onload = () => {
        window.TamaraProductWidget.init({ lang: this.language });
        setTimeout(() => {
          window.TamaraProductWidget.render();
        }, 300);
      };
    }
    // tabby
    if (this.tabbyIsActive) {
      var script = document.createElement('script');
      script.setAttribute('src', 'https://checkout.tabby.ai/tabby-promo.js');
      document.head.appendChild(script);
      script.onload = () => {
        const TabbyPromo = window.TabbyPromo;
        new TabbyPromo({
          selector: '#tabbyPromo',
          currency: this.currency,
          price: this.price,
          lang: this.language,
        });
        document.querySelectorAll('.tabby-promo-snippet__logo').forEach(function (element) {
          element.setAttribute('aria-label', 'Tabby Logo');
        });
      };
      //this is workaround to remove the default border and add margin
      this.removeTabbyBorder();
    }
    // Spotii
    if (this.spotiiIsActive) {
      let amount = salla.money((Number(this.price) / 3).toFixed(2));
      let isRTL = salla.config.get('theme.is_rtl', true);
      window.spotiiConfig = {
        targetXPath: ['.spotii-wrapper'],
        renderToPath: ['.spotii-promo'],
        numberOfPayment: 3,
        currency: this.currency,
        templateLine: "${textOne} ${number} ${textTwo} " + amount + "${logo} ${info}",
        //todo:: translate these
        textOne: isRTL ? "جزء الدفع على" : "Split it into",
        textTwo: isRTL ? "أقساط متساوية بدون تكاليف اضافية بقيمة" : "payments of",
        textThree: "مع",
        price: this.price,
        // forcedShow: false,
        // merchantID: null,
      };
      var script = document.createElement('script');
      script.setAttribute('src', salla.url.cdn('js/price-widget-ar-salla.js'));
      document.head.appendChild(script);
      // script.onload = () => {
      //   // setTimeout()
      // }
    }
  }
  /**
   * this is workaround to remove the default border and add margin
   * we will try to remove tabby border 5 times for 7.5 seconds
   */
  removeTabbyBorder() {
    if (this.tabbyBorderRemoved || this.tabbyRemoveBorderTries > 5) {
      return;
    }
    this.tabbyRemoveBorderTries++;
    setTimeout(() => {
      let promo = document.querySelector('#tabbyPromo>div>div');
      promo = promo ? promo.shadowRoot.querySelector('div[class^="styles__tabby-promo-snippet--"]') : null;
      if (promo) {
        promo.style = 'border: none; margin: 15px 0!important;';
        this.tabbyBorderRemoved = true;
      }
      else {
        this.removeTabbyBorder();
      }
    }, this.tabbyRemoveBorderTries * 500);
  }
  static get is() { return "salla-installment"; }
  static get originalStyleUrls() {
    return {
      "$": ["salla-installment.scss"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["salla-installment.css"]
    };
  }
  static get properties() {
    return {
      "price": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Current product price"
        },
        "attribute": "price",
        "reflect": false
      },
      "language": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Language code"
        },
        "attribute": "language",
        "reflect": false,
        "defaultValue": "salla.config.get('user.language_code')"
      },
      "currency": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Currency code"
        },
        "attribute": "currency",
        "reflect": false,
        "defaultValue": "salla.config.get('user.currency_code')"
      }
    };
  }
  static get states() {
    return {
      "tamaraIsActive": {},
      "tabbyIsActive": {},
      "spotiiIsActive": {}
    };
  }
}
