// @salla.sa/twilight-components/dist/collection/components/salla-placeholder/salla-placeholder.js
/*!
 * Crafted with â¤ by Salla
 */
import { h, Host } from '@stencil/core';
import Inbox from '../../assets/svg/inbox.svg';
/**
 * @slot title - The primary content of the placeholder.
 * @slot description - Additional content displayed below the title.
 */
export class SallaPlaceholder {
  constructor() {
    this.translationLoaded = false;
    this.icon = Inbox;
    this.alignment = 'left';
    this.iconSize = 'md';
    salla.lang.onLoaded(() => {
      this.translationLoaded = true;
    });
  }
  alignmentClass() {
    return {
      's-placeholder-wrapper': true,
      's-placeholder-align-left': this.alignment == 'left',
      's-placeholder-align-center': this.alignment == 'center',
      's-placeholder-align-right': this.alignment == 'right',
    };
  }
  render() {
    return (h(Host, { class: this.alignmentClass() }, h("div", { class: `s-placeholder-icon s-placeholder-icon-${this.iconSize}`, innerHTML: this.icon }), h("div", { class: "s-placeholder-title" }, h("slot", { name: 'title' }, h("span", null, salla.lang.get('common.elements.no_options')))), h("div", { class: "s-placeholder-description" }, h("slot", { name: 'description' }, h("span", null, salla.lang.get('common.errors.empty_results'))))));
  }
  static get is() { return "salla-placeholder"; }
  static get originalStyleUrls() {
    return {
      "$": ["salla-placeholder.scss"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["salla-placeholder.css"]
    };
  }
  static get properties() {
    return {
      "icon": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Custom icon to display, defaults to [`Inbox`], can be any valid svg icon or inline element with font icon"
        },
        "attribute": "icon",
        "reflect": false,
        "defaultValue": "Inbox"
      },
      "alignment": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "'left' | 'center' | 'right'",
          "resolved": "\"center\" | \"left\" | \"right\"",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Defines the alignment of contents. Defaults to [`left`]"
        },
        "attribute": "alignment",
        "reflect": false,
        "defaultValue": "'left'"
      },
      "iconSize": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "'xs' | 'sm' | 'md' | 'lg' | 'xl' | 'xxl' | 'xxxl'",
          "resolved": "\"lg\" | \"md\" | \"sm\" | \"xl\" | \"xs\" | \"xxl\" | \"xxxl\"",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "The size of the icon. Defaults to [`md`] = 45px for font icon & width/height: 3.5rem for svg icon"
        },
        "attribute": "icon-size",
        "reflect": false,
        "defaultValue": "'md'"
      }
    };
  }
  static get states() {
    return {
      "translationLoaded": {}
    };
  }
}
