// @salla.sa/twilight-components/dist/collection/components/salla-rating-modal/salla-rating-modal.js
/*!
 * Crafted with â¤ by Salla
 */
import { h, Host } from '@stencil/core';
import Star from '../../assets/svg/star.svg';
import ShippingFast from "../../assets/svg/shipping-fast.svg";
import CheckCircle2 from "../../assets/svg/check-circle2.svg";
import Helper from '../../Helpers/Helper';
export class SallaRatingModal {
  constructor() {
    this.stepsCount = 0;
    this.currentIndex = 0;
    this.submitted = [];
    this.order = undefined;
    this.hasError = false;
    this.errorMessage = undefined;
    this.orderId = salla.config.get('page.id');
    this.translationLoaded = false;
    salla.event.on('rating::open', () => this.open());
    salla.lang.onLoaded(() => {
      this.translationLoaded = true;
    });
  }
  /**
   * Show the rating modal
   */
  async open() {
    return this.modal.open()
      .then(() => this.order || salla.api.withoutNotifier(() => salla.rating.api.order(this.orderId)).then(res => this.order = res.data))
      .then(() => this.modal.setTitle(salla.lang.get('pages.rating.rate_order') + ' <span class="unicode">(#' + this.order.id + ')</span>'))
      .then(() => this.modal.stopLoading())
      .then(() => this.stepsCount = [this.order.testimonials_enabled, this.order.products_enabled, this.order.shipping_enabled].filter(enabled => enabled).length)
      .then(() => setTimeout(() => this.handleWizard(), 100))
      .catch(e => {
      var _a, _b, _c, _d;
      this.hasError = true;
      this.errorMessage = ((_c = (_b = (_a = e.response) === null || _a === void 0 ? void 0 : _a.data) === null || _b === void 0 ? void 0 : _b.error) === null || _c === void 0 ? void 0 : _c.message) || ((_d = e.response) === null || _d === void 0 ? void 0 : _d.data);
      this.modal.stopLoading();
    });
  }
  /**
   * Show the rating modal
   */
  async close() {
    return this.modal.close();
  }
  // handle wizard
  handleWizard() {
    this.steps = this.modal.querySelectorAll(".s-rating-modal-step");
    this.dots = this.modal.querySelectorAll(".s-rating-modal-step-dot");
    this.showActiveStep();
  }
  showActiveStep(current = null) {
    var _a, _b;
    this.currentTab = current || this.steps[this.currentIndex];
    Helper.toggleClassIf('.s-rating-modal-step-dot', 's-rating-modal-bg-gray', 's-rating-modal-bg-primary', dot => dot != this.dots[this.currentIndex])
      .toggleClassIf('.s-rating-modal-step', 's-rating-modal-active', 's-rating-modal-hidden', tab => tab == this.currentTab);
    if (this.currentIndex != 0) {
      // the animation
      Helper.toggleElementClassIf(this.currentTab, 's-rating-modal-unactive', 's-rating-modal-hidden', () => true);
      setTimeout(() => Helper.toggleElementClassIf(this.currentTab, 's-rating-modal-active', 's-rating-modal-unactive', () => true), 300);
    }
    // Btn text
    let nextType = (_a = this.steps[this.currentIndex + 1]) === null || _a === void 0 ? void 0 : _a.dataset.type;
    (_b = this.nextBtn) === null || _b === void 0 ? void 0 : _b.setText(nextType ? salla.lang.get('pages.rating.rate') + ' ' + salla.lang.get('pages.rating.' + nextType)
      : salla.lang.get('pages.rating.send_ratings'));
    setTimeout(() => { var _a, _b; return (_a = this.body) === null || _a === void 0 ? void 0 : _a.setAttribute('style', 'height:' + ((_b = this.currentTab) === null || _b === void 0 ? void 0 : _b.offsetHeight) + 'px'); });
  }
  previousTab() {
    this.currentIndex > 0 && this.currentIndex--;
    Helper.toggleElementClassIf(this.backBtn, 's-rating-modal-unvisiable', 'block', () => this.currentIndex == 0);
    this.showActiveStep();
  }
  submit() {
    this.submittedBefore() || this.validate();
    salla.config.canLeave = false;
    this.nextBtn.load()
      .then(() => this.submittedBefore() || this.sendFeedback())
      .then(() => this.currentTab.querySelectorAll('[name],.s-rating-modal-btn-star').forEach(el => el.setAttribute('disabled', '')))
      .then(() => this.currentIndex < this.stepsCount && this.currentIndex++)
      .then(() => this.showActiveStep())
      .then(() => Helper.toggleClassIf('#prev-btn', 'block', 's-rating-modal-unvisiable', () => true))
      .finally(() => {
      this.nextBtn.stop();
      salla.config.canLeave = true;
      this.currentIndex == this.stepsCount && this.showThankYou();
      this.modal.isClosable = false;
    });
  }
  submittedBefore() {
    return this.submitted.includes(this.currentIndex);
  }
  validate(rating = null, type = null) {
    if (!rating && this.currentTab.dataset.type == 'products') {
      return this.currentTab.querySelectorAll('.rating-outer-form').forEach(rating => this.validate(rating, 'product'));
    }
    rating = rating || this.currentTab;
    let stars = rating.querySelector('.rating_hidden_input').value;
    let comment = rating.querySelector('.s-rating-modal-comment');
    let validationMessage = rating.querySelector('.s-rating-modal-validation-msg');
    if (stars && comment.value && comment.value.length > 3) {
      comment.classList.remove('s-has-error');
      validationMessage.innerHTML = '';
      return;
    }
    type = type || rating['dataset'].type;
    Helper.toggleElementClassIf(comment, 'save', 's-has-error', el => el.value.length > 3);
    throw validationMessage.innerHTML = stars
      ? (salla.lang.get('common.errors.not_less_than_chars', { chars: 4 }) + ' ' + comment.getAttribute('placeholder'))
      : salla.lang.get(`pages.rating.rate_${type}_stars`).replace(' (:item)', '');
  }
  sendFeedback() {
    let data = {};
    this.currentTab.querySelectorAll('[name]').forEach((input) => {
      //decode names like `<input name="jamal[inner]" value="hi">` to be {name:jamal, value: {inner:"hi"}}
      let inputData = salla.helpers.inputData(input.name, input.value, data);
      data[inputData.name] = inputData.value;
    });
    if (Object.keys(data).length == 0) {
      return;
    }
    data['order_id'] = this.orderId;
    data['type'] = this.currentTab.dataset.type;
    this.submitted.push(this.currentIndex);
    return salla.rating.api[this.currentTab.dataset.type](data);
  }
  showThankYou() {
    let seconds = 10;
    let timeToClose = setInterval(() => {
      this.thanksTime.innerHTML = '00:0' + (seconds--);
      if (seconds > 0) {
        return;
      }
      clearInterval(timeToClose);
      this.thanksTime.remove();
      this.close().then(() => window.location.reload());
    }, 1000);
    this.modal.querySelector('.s-rating-modal-footer').classList.add('s-rating-modal-unvisiable');
    this.showActiveStep(this.thanksTab);
  }
  render() {
    var _a;
    return (h(Host, { class: "s-rating-modal" }, h("salla-modal", { class: "s-rating-modal-wrap", isLoading: true, width: "md", ref: modal => this.modal = modal }, h("div", { slot: 'loading' }, h("div", { class: "s-rating-modal-skeleton" }, h("salla-skeleton", { type: 'circle', height: '80px', width: '80px' }), h("salla-skeleton", { height: '15px', width: '60%' }), h("salla-skeleton", { height: '10px', width: '30%' }), h("div", { class: "s-rating-modal-skeleton-stars" }, [...Array(5)].map(() => h("div", { innerHTML: Star }))), h("salla-skeleton", { height: '100px', width: '100%' }), h("div", { class: "s-rating-modal-skeleton-footer" }, h("salla-skeleton", { height: '40px', width: '30%' })))), !this.hasError && this.order
      ? [h("div", { class: "s-rating-modal-wrapper", ref: el => this.body = el }, ((_a = this.order) === null || _a === void 0 ? void 0 : _a.testimonials_enabled) ?
          h("div", { class: "rating-outer-form s-rating-modal-step-wrap s-rating-modal-step s-rating-modal-hidden", "data-type": "store" }, h("div", { class: "s-rating-modal-rounded-icon" }, h("img", { src: salla.config.get('store.logo', 'https://assets.salla.sa/cp/assets/images/logo-new.png'), alt: "store name", class: "s-rating-modal-store-logo" })), h("h2", { class: "s-rating-modal-title" }, salla.lang.get('pages.rating.rate_the_store')), h("div", { class: "s-rating-modal-stars-company" }, h("salla-rating-stars", { size: "large" })), h("textarea", { name: "comment", class: "s-rating-modal-comment", placeholder: salla.lang.get('pages.rating.write_store_rate') }), h("small", { class: "s-rating-modal-validation-msg" }))
          : '', this.order.products_enabled
          ? h("section", { class: "s-rating-modal-step s-rating-modal-hidden", "data-type": "products" }, this.order.products.map((item, index) => h("div", { class: "rating-outer-form s-rating-modal-product", "data-stars-error": salla.lang.get('pages.rating.rate_product_stars') }, h("div", { class: "s-rating-modal-product-img-wrap" }, h("img", { src: item.product.thumbnail, alt: item.product.name, class: "s-rating-modal-product-img" })), h("div", { class: "s-rating-modal-product-details" }, h("h3", { class: "s-rating-modal-product-title" }, " ", item.product.name), h("div", { class: "s-rating-modal-stars-product" }, h("salla-rating-stars", { size: "small", name: `products[${index}][rating]` })), h("input", { type: "hidden", name: `products[${index}][product_id]`, value: item.product.id }), h("textarea", { placeholder: salla.lang.get('pages.rating.write_product_rate'), name: `products[${index}][comment]`, class: "s-rating-modal-comment" }), h("small", { class: "s-rating-modal-validation-msg" })))))
          : '', this.order.shipping_enabled
          ? h("div", { class: "rating-outer-form s-rating-modal-step-wrap s-rating-modal-step s-rating-modal-hidden", "data-type": "shipping" }, h("input", { type: "hidden", name: "shipping_company_id", value: this.order.shipping.company.id }), this.order.shipping.company.logo
            ? h("div", { class: "s-rating-modal-rounded-icon" }, h("img", { src: this.order.shipping.company.logo, class: "s-rating-modal-shipping-logo", alt: this.order.shipping.company.name }))
            : h("span", { class: "s-rating-modal-icon", innerHTML: ShippingFast }), h("div", { class: "s-rating-modal-title" }, " ", salla.lang.get('pages.rating.rate_shipping') + ' ' + this.order.shipping.company.name), h("div", { class: "s-rating-modal-stars-company" }, h("salla-rating-stars", { size: "large" })), h("textarea", { name: "comment", class: "s-rating-modal-comment", placeholder: salla.lang.get('pages.rating.write_shipping_rate') }), h("small", { class: "s-rating-modal-validation-msg" }))
          : '', h("div", { class: "s-rating-modal-thanks s-rating-modal-hidden", ref: el => this.thanksTab = el }, h("span", { class: "s-rating-modal-icon", innerHTML: CheckCircle2 }), h("h3", { class: "s-rating-modal-thanks-title" }, salla.lang.get('pages.rating.thanks')), h("div", { class: "s-rating-modal-thanks-msg", innerHTML: this.order.thanks_message }), h("time", { class: "s-rating-modal-thanks-time", ref: el => this.thanksTime = el }))), h("div", { class: "s-rating-modal-footer" }, h("button", { ref: el => this.backBtn = el, onClick: () => this.previousTab(), class: "s-rating-modal-btn s-rating-modal-unvisiable" }, salla.lang.get('common.elements.back')), this.stepsCount > 1 ? h("ul", { class: "s-rating-modal-dots" }, [0, 1, 2].slice(0, this.stepsCount).map(() => h("li", { class: 's-rating-modal-bg-gray s-rating-modal-step-dot' }))) : '', h("salla-button", { "loader-position": 'center', ref: el => this.nextBtn = el, onClick: () => this.submit() }, salla.lang.get('common.elements.next'))),]
      : h("salla-placeholder", { alignment: "center" }))));
  }
  componentDidLoad() {
    salla.event.dispatch('rating::ready', this);
  }
  static get is() { return "salla-rating-modal"; }
  static get originalStyleUrls() {
    return {
      "$": ["salla-rating-modal.scss"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["salla-rating-modal.css"]
    };
  }
  static get properties() {
    return {
      "orderId": {
        "type": "number",
        "mutable": false,
        "complexType": {
          "original": "number",
          "resolved": "number",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "The order id, to rate on its products & shipping"
        },
        "attribute": "order-id",
        "reflect": false,
        "defaultValue": "salla.config.get('page.id')"
      }
    };
  }
  static get states() {
    return {
      "order": {},
      "hasError": {},
      "errorMessage": {},
      "translationLoaded": {}
    };
  }
  static get methods() {
    return {
      "open": {
        "complexType": {
          "signature": "() => Promise<void | NodeJS.Timeout>",
          "parameters": [],
          "references": {
            "Promise": {
              "location": "global"
            },
            "NodeJS": {
              "location": "global"
            }
          },
          "return": "Promise<void | Timeout>"
        },
        "docs": {
          "text": "Show the rating modal",
          "tags": []
        }
      },
      "close": {
        "complexType": {
          "signature": "() => Promise<HTMLElement>",
          "parameters": [],
          "references": {
            "Promise": {
              "location": "global"
            },
            "HTMLElement": {
              "location": "global"
            }
          },
          "return": "Promise<HTMLElement>"
        },
        "docs": {
          "text": "Show the rating modal",
          "tags": []
        }
      }
    };
  }
  static get elementRef() { return "host"; }
}
