// @salla.sa/twilight-components/dist/collection/components/salla-tabs/salla-tab-content.js
/*!
 * Crafted with â¤ by Salla
 */
import { h } from '@stencil/core';
/**
 * @slot The default slot.
 */
export class SallaTabContent {
  constructor() {
    this.name = undefined;
    this.isSelected = false;
  }
  /**
   * Expose self for the parent.
   */
  async getChild() {
    return {
      selected: this.selected.bind(this),
      unselect: this.unselect.bind(this),
      name: this.name
    };
  }
  unselect() {
    this.isSelected = false;
  }
  selected() {
    this.isSelected = true;
  }
  render() {
    const classes = {
      's-tabs-content': true,
      's-tabs-content-selected': this.isSelected
    };
    return (h("div", { class: classes }, h("slot", null)));
  }
  static get is() { return "salla-tab-content"; }
  static get originalStyleUrls() {
    return {
      "$": ["salla-tab-content.scss"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["salla-tab-content.css"]
    };
  }
  static get properties() {
    return {
      "name": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Set name of the tab content. Mainly used as a key to s\nynchronize the content with it's respective header."
        },
        "attribute": "name",
        "reflect": false
      }
    };
  }
  static get states() {
    return {
      "isSelected": {}
    };
  }
  static get methods() {
    return {
      "getChild": {
        "complexType": {
          "signature": "() => Promise<{ selected: any; unselect: any; name: string; }>",
          "parameters": [],
          "references": {
            "Promise": {
              "location": "global"
            }
          },
          "return": "Promise<{ selected: any; unselect: any; name: string; }>"
        },
        "docs": {
          "text": "Expose self for the parent.",
          "tags": []
        }
      }
    };
  }
}
