// @salla.sa/twilight-components/dist/collection/components/salla-social-share/salla-social-share.js
/*!
 * Crafted with â¤ by Salla
 */
import { h } from '@stencil/core';
import anime from 'animejs';
import facebook from '../../assets/svg/facebook.svg';
import twitter from '../../assets/svg/twitter.svg';
import copy_link from '../../assets/svg/link.svg';
import email from '../../assets/svg/mail.svg';
import whatsapp from '../../assets/svg/whatsapp.svg';
import { Socials } from './interfaces';
import Cancel from "../../assets/svg/cancel.svg";
import ShareAlt from "../../assets/svg/share-alt.svg";
/**
 * @slot widget - An action that can be used to activete or open the component by calling the `open` method.
 */
export class SallaSocialShare {
  constructor() {
    this.opened = false;
    this.allPlatforms = [Socials.WHATSAPP, Socials.FACEBOOK, Socials.TWITTER, Socials.EMAIL, Socials.COPY_LINK];
    this.platformIcons = [{ icon: whatsapp, name: Socials.WHATSAPP }, {
        icon: facebook,
        name: Socials.FACEBOOK
      }, { icon: twitter, name: Socials.TWITTER }, { icon: email, name: Socials.EMAIL }, {
        icon: copy_link,
        name: Socials.COPY_LINK
      }];
    this.convertedPlatforms = [];
    this.url = "";
    this.urlName = "";
    this.platforms = 'facebook,twitter,whatsapp,email,copy_link';
  }
  componentWillLoad() {
    let that = this;
    if (document.getElementById('a2a-script')) {
      return;
    }
    const script = document.createElement('script');
    script.src = 'https://static.addtoany.com/menu/page.js';
    script.setAttribute('id', 'a2a-script');
    script.async = true;
    script.onload = function () {
      // ? this will not work if there is more than one share component in the page
      window.a2a_config = window.a2a_config || {};
      window.a2a_config.linkurl = that.url ? that.url : window.location.href;
      window.a2a_config.linkname = that.urlName ? that.urlName : document.title;
      window.a2a_config.locale = salla.config.get('user.language_code', salla.lang.getLocale());
    };
    document.body.appendChild(script);
  }
  /**
   * Activate or open the share menu.
   */
  async open() {
    this.opened = !this.opened;
    if (this.opened) {
      this.shareMenu.classList.add('opened');
    }
    else {
      this.shareMenu.classList.remove('opened');
    }
    new anime.timeline().add({
      targets: this.shareMenu,
      translateY: [-50, 0],
      opacity: [0, 1],
      duration: 300,
      podding: '0',
      easing: 'easeInOutSine'
    }).add({
      targets: this.platformItem,
      translateZ: 0,
      translateY: [-30, 0],
      scaleY: [0, 1],
      opacity: [0, 1],
      duration: 1200,
      delay: anime.stagger(100)
    }, '-=200');
  }
  render() {
    return (h("div", { class: "s-social-share-wrapper" }, h("slot", { name: 'widget' }, h("salla-button", { "aria-label": "Share", onClick: () => this.open(), class: "s-social-share-btn", shape: "icon", fill: "outline", color: "light" }, h("span", { innerHTML: this.opened ? Cancel : ShareAlt }))), h("ul", { ref: el => this.shareMenu = el, class: "s-social-share-list a2a_kit share" }, this.platforms.split(',').map(platform => {
      return (h("li", { ref: el => this.platformItem = el }, h("a", { class: `a2a_button_${platform}`, "aria-label": `Share Via ${platform}` }, this.platformIcons.map((icon) => {
        if (icon.name === platform) {
          return h("span", { class: "s-social-share-icon", innerHTML: icon.icon });
        }
      }))));
    }))));
  }
  static get is() { return "salla-social-share"; }
  static get originalStyleUrls() {
    return {
      "$": ["salla-social-share.scss"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["salla-social-share.css"]
    };
  }
  static get properties() {
    return {
      "url": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "page url that will be shared custom | current page url"
        },
        "attribute": "url",
        "reflect": true,
        "defaultValue": "\"\""
      },
      "urlName": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "page url name that will be shared custom | current page url"
        },
        "attribute": "url-name",
        "reflect": true,
        "defaultValue": "\"\""
      },
      "platforms": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "selected platforms to share | all platforms"
        },
        "attribute": "platforms",
        "reflect": true,
        "defaultValue": "'facebook,twitter,whatsapp,email,copy_link'"
      }
    };
  }
  static get states() {
    return {
      "opened": {},
      "allPlatforms": {},
      "platformIcons": {},
      "convertedPlatforms": {}
    };
  }
  static get methods() {
    return {
      "open": {
        "complexType": {
          "signature": "() => Promise<void>",
          "parameters": [],
          "references": {
            "Promise": {
              "location": "global"
            }
          },
          "return": "Promise<void>"
        },
        "docs": {
          "text": "Activate or open the share menu.",
          "tags": []
        }
      }
    };
  }
}
