// @salla.sa/twilight-components/dist/collection/components/salla-price-range/salla-price-range.js
/*!
 * Crafted with ❤ by Salla
 */
import { Host, h } from '@stencil/core';
export class SallaPriceRange {
  constructor() {
    this.minPrice = 1000;
    this.maxPrice = 5000;
    this.option = undefined;
    this.filtersData = undefined;
    this.min = 0;
    this.max = 10000;
    this.minthumb = 0;
    this.maxthumb = 0;
    this.priceOptions = undefined;
    this.generatePriceOptions();
  }
  async reset() {
    var _a, _b;
    //@ts-ignore
    this.minInput.value = `${((_b = (_a = this.option) === null || _a === void 0 ? void 0 : _a.values) === null || _b === void 0 ? void 0 : _b.min) || 0}`;
    this.maxInput.value = `${this.max}`;
  }
  generatePriceOptions() {
    const MAX_PRICE = this.max;
    let option1 = Math.floor(MAX_PRICE / 100);
    if (option1 < 1) {
      option1 = Math.floor(MAX_PRICE / 10);
    }
    let option2 = Math.floor(MAX_PRICE / 2);
    let option3 = Math.floor((MAX_PRICE / 3) * 2);
    this.priceOptions = [option1, option2, option3];
    // to be removed
    console.log(`Price options for max price ${MAX_PRICE}: ${this.priceOptions.join(', ')}`);
  }
  getPriceLabel(price) {
    if (price == Math.floor(this.max / 100)) {
      return `اقل من ${salla.money(price)}`;
    }
    else if (price == Math.floor(this.max / 2)) {
      return `${salla.money(this.priceOptions[0])} الى ${salla.money(this.priceOptions[2])}`;
    }
    else {
      return `أكثر من ${salla.money(price)}`;
    }
  }
  handleMinMaxPrice(event, option) {
    const [firstOption, secondOption, thirdOption] = this.priceOptions;
    switch (option) {
      case firstOption:
        this.minPrice = 0;
        this.maxPrice = firstOption;
        break;
      case secondOption:
        this.minPrice = firstOption;
        this.maxPrice = thirdOption;
        break;
      case thirdOption:
        this.minPrice = thirdOption;
        this.maxPrice = this.max;
        break;
      default:
        throw new Error(`Invalid price option: ${option}`);
    }
    // Update the input values
    this.minInput.value = `${this.minPrice}`;
    this.maxInput.value = `${this.maxPrice}`;
    this.changedEventHandler(event);
  }
  async changedEventHandler(event, isMin = false) {
    let value = event ? event.target.value * 1 : null;
    if (isMin) {
      this.minInputValidation(value);
      this.minthumb = ((this.minPrice - this.min) / (this.max - this.min)) * 100;
    }
    else {
      this.maxInputValidation(value);
      this.maxthumb = 100 - (((this.maxPrice - this.min) / (this.max - this.min)) * 100);
    }
    this.isReady && this.changed.emit({
      event: event,
      option: this.option,
      value: { max: this.maxPrice, min: this.minPrice }
    });
  }
  minInputValidation(value) {
    if (value && (value > this.max || value > this.maxPrice)) {
      // this.minPrice = this.maxPrice;
      return;
    }
    if (value < this.min) {
      this.minPrice = this.min;
      return;
    }
    if (value) {
      this.minPrice = value;
    }
  }
  maxInputValidation(value) {
    if (value && (value < this.min || value < this.minPrice)) {
      // this.maxPrice = this.minPrice;
      return;
    }
    if (value > this.max) {
      this.maxPrice = this.max;
      return;
    }
    if (value) {
      this.maxPrice = value;
    }
  }
  isChecked(option) {
    if (option == Math.floor(this.max / 100)) {
      return this.minPrice == 0 && this.maxPrice == option;
    }
    if (option == Math.floor(this.max / 2)) {
      return this.minPrice == this.priceOptions[0] && this.maxPrice == this.priceOptions[2];
    }
    if (option == Math.floor((this.max / 3) * 2)) {
      return this.minPrice == option && this.maxPrice == this.max;
    }
  }
  render() {
    return (h(Host, null, this.priceOptions && this.priceOptions.map((option, index) => {
      return h("label", { class: "s-filters-label", htmlFor: `${option}-${index}` }, h("input", { id: `${option}-${index}`, name: "price", type: "radio", checked: this.isChecked(option), class: "s-filters-radio", onChange: e => this.handleMinMaxPrice(e, option) }), this.getPriceLabel(option));
    }), h("div", { class: "flex justify-center items-center" }, h("div", { class: "relative max-w-xl w-full" }, h("div", { class: "s-price-range-inputs" }, h("div", { class: "s-price-range-relative" }, h("div", { class: "s-price-range-currency" }, " ", salla.config.currency().symbol), h("input", { type: "number", maxlength: "5", ref: el => this.minInput = el, onInput: (event) => this.changedEventHandler(event, true),
      // value={this.minPrice}
      placeholder: "\u0645\u0646", class: "s-price-range-number-input" })), h("div", { class: "s-price-range-gray-text" }, " -"), h("div", { class: "s-price-range-relative" }, h("div", { class: "s-price-range-currency" }, " ", salla.config.currency().symbol), h("input", { type: "number", maxlength: "5", placeholder: "\u0627\u0644\u0649", ref: el => this.maxInput = el, onInput: (event) => this.changedEventHandler(event),
      // value={this.maxPrice}
      class: "s-price-range-number-input", "aria-describedby": "price-currency" })))))));
  }
  componentDidLoad() {
    var _a;
    if (this.filtersData && ((_a = this.filtersData) === null || _a === void 0 ? void 0 : _a.price)) {
      this.minPrice = this.filtersData.price.min;
      this.maxPrice = this.filtersData.price.max;
      this.maxInput.value = `${this.maxPrice}`;
      this.minInput.value = `${this.minPrice}`;
    }
    this.isReady = true;
  }
  static get is() { return "salla-price-range"; }
  static get originalStyleUrls() {
    return {
      "$": ["salla-price-range.scss"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["salla-price-range.css"]
    };
  }
  static get properties() {
    return {
      "minPrice": {
        "type": "number",
        "mutable": true,
        "complexType": {
          "original": "number",
          "resolved": "number",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "min-price",
        "reflect": false,
        "defaultValue": "1000"
      },
      "maxPrice": {
        "type": "number",
        "mutable": true,
        "complexType": {
          "original": "number",
          "resolved": "number",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "max-price",
        "reflect": false,
        "defaultValue": "5000"
      },
      "option": {
        "type": "unknown",
        "mutable": false,
        "complexType": {
          "original": "Filter",
          "resolved": "Filter",
          "references": {
            "Filter": {
              "location": "import",
              "path": "../salla-filters/interfaces"
            }
          }
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        }
      },
      "filtersData": {
        "type": "any",
        "mutable": false,
        "complexType": {
          "original": "any",
          "resolved": "any",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "filters-data",
        "reflect": true
      }
    };
  }
  static get states() {
    return {
      "min": {},
      "max": {},
      "minthumb": {},
      "maxthumb": {},
      "priceOptions": {}
    };
  }
  static get events() {
    return [{
        "method": "changed",
        "name": "changed",
        "bubbles": true,
        "cancelable": true,
        "composed": true,
        "docs": {
          "tags": [],
          "text": ""
        },
        "complexType": {
          "original": "any",
          "resolved": "any",
          "references": {}
        }
      }];
  }
  static get methods() {
    return {
      "reset": {
        "complexType": {
          "signature": "() => Promise<void>",
          "parameters": [],
          "references": {
            "Promise": {
              "location": "global"
            }
          },
          "return": "Promise<void>"
        },
        "docs": {
          "text": "",
          "tags": []
        }
      }
    };
  }
}
