// @salla.sa/twilight-components/dist/collection/components/salla-search/salla-search.js
/*!
 * Crafted with â¤ by Salla
 */
import { h } from '@stencil/core';
import Search from "../../assets/svg/search.svg";
import Helper from '../../Helpers/Helper';
/**
 * @slot product - Replaces products card in the results, has replaceable props `{name}`, `{price}`, `{regular_price}`, `{image}`.
 */
export class SallaSearch {
  constructor() {
    var _a;
    this.inputValue = '';
    this.translationLoaded = false;
    this.results = undefined;
    this.loading = false;
    this.typing = false;
    this.debounce = setTimeout(() => '', 1000);
    this.search_term = undefined;
    this.inline = false;
    this.oval = false;
    this.height = 60;
    this.productSlot = ((_a = this.host.querySelector('[slot="product"]')) === null || _a === void 0 ? void 0 : _a.innerHTML) || this.getDefaultProductSlot();
    salla.event.on('search::open', () => this.open());
    salla.lang.onLoaded(() => {
      this.translationLoaded = true;
    });
    salla.event.on('modalClosed', () => this.onModalClose());
  }
  async open() {
    if (!this.inline) {
      await this.modal.open().then(() => setTimeout(() => this.searchInput.focus(), 300));
    }
  }
  onModalClose() {
    this.searchInput.value = '';
    this.results = undefined;
    this.afterSearching();
    this.container.classList.remove('s-search-no-results');
  }
  handleKeyDown(ev) {
    var _a;
    if (ev.key === 'Enter' && ((_a = this.search_term) === null || _a === void 0 ? void 0 : _a.length)) {
      window.location.href = salla.url.get('search?q=' + encodeURI(this.search_term));
    }
  }
  getDefaultProductSlot() {
    return '<div class="s-search-product-image-container">' +
      '  <img class="s-search-product-image" src="{image}" alt="{name}"/>' +
      '</div>' +
      '<div class="s-search-product-details">' +
      '  <div class="s-search-product-title">{name}</div> <div class="s-search-product-price">{price} <span class="s-search-product-regular-price">{regular_price}</span></div>' +
      '</div>';
  }
  debounceSearch(event) {
    this.typing = true;
    clearTimeout(this.debounce);
    this.debounce = setTimeout(() => {
      this.typing = false;
      this.search_term = event.target.value;
    }, 700);
  }
  handleSearch(val) {
    this.inputValue = val;
    if (val.length > 2) {
      this.search(val);
    }
    else {
      this.results = undefined;
      this.afterSearching();
    }
  }
  search(val) {
    this.noResults.style.display = 'none';
    //run loading spinner or stop it
    this.loading = true;
    salla.product.search(val)
      .then(response => this.results = response)
      .catch(err => err !== 'Query is same as previous one!' ? this.results = undefined : null)
      .finally(() => this.afterSearching(/*isEmpty*/ false));
  }
  afterSearching(isEmpty = true) {
    var _a;
    this.noResults.style.display = isEmpty || ((_a = this.results) === null || _a === void 0 ? void 0 : _a.data.length) > 0 ? 'none' : 'block';
    Helper.toggleElementClassIf(this.container, 's-search-container-open', 's-search-no-results', () => { var _a; return (_a = this.results) === null || _a === void 0 ? void 0 : _a.data.length; });
    this.loading = false;
    salla.product.api.previousQuery = ''; //avoid having error 'Query is same as previous one!' after reopen modal;
    this.inputValue.length < 3 ? this.container.classList.remove('s-search-no-results') : '';
  }
  render() {
    var _a;
    const searchContent = h("div", { class: { 's-search-container': true, 's-search-inline': this.inline }, ref: container => this.container = container }, h("input", { type: "search", enterkeyhint: "search", autocomplete: "off", class: "s-search-input", placeholder: salla.lang.get('blocks.header.search_placeholder'), onInput: e => this.debounceSearch(e), onKeyDown: e => this.handleKeyDown(e), ref: input => this.searchInput = input, style: { height: this.height + 'px', borderRadius: this.oval ? this.height / 2 + 'px' : '' } }), h("span", { class: "s-search-icon-wrap" }, h("span", { class: "s-search-icon", innerHTML: this.loading ? '<i class="s-search-spinner-loader"></i>' : Search })), h("div", { class: "s-search-results" }, (_a = this.results) === null || _a === void 0 ? void 0 :
      _a.data.map(item => h("a", { href: item.url, class: { "s-search-product": true, 's-search-product-not-available': !item.is_available }, innerHTML: this.productSlot
          .replace(/\{name\}/g, item.name)
          .replace(/\{price\}/g, salla.money(item.price))
          .replace(/\{regular_price\}/g, item.has_special_price ? salla.money(item.regular_price) : '')
          .replace(/\{image\}/g, item.thumbnail) })), h("p", { ref: el => this.noResults = el, class: "s-search-no-results-placeholder" }, salla.lang.get('common.elements.no_options'))));
    return (this.inline ?
      h("div", { class: "s-search-modal" }, searchContent)
      :
        h("salla-modal", { position: "top", class: "s-search-modal", ref: modal => this.modal = modal }, searchContent));
  }
  /**
   * Run it one time after load
   */
  componentDidLoad() {
    this.afterSearching();
  }
  static get is() { return "salla-search"; }
  static get originalStyleUrls() {
    return {
      "$": ["salla-search.scss"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["salla-search.css"]
    };
  }
  static get properties() {
    return {
      "inline": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Set the component display without modal window. Defaults to `false`"
        },
        "attribute": "inline",
        "reflect": false,
        "defaultValue": "false"
      },
      "oval": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Adds a border radius to the input. Half of the height."
        },
        "attribute": "oval",
        "reflect": false,
        "defaultValue": "false"
      },
      "height": {
        "type": "number",
        "mutable": false,
        "complexType": {
          "original": "number",
          "resolved": "number",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Sets the height of the input"
        },
        "attribute": "height",
        "reflect": false,
        "defaultValue": "60"
      }
    };
  }
  static get states() {
    return {
      "translationLoaded": {},
      "results": {},
      "loading": {},
      "typing": {},
      "debounce": {},
      "search_term": {}
    };
  }
  static get elementRef() { return "host"; }
  static get watchers() {
    return [{
        "propName": "search_term",
        "methodName": "handleSearch"
      }];
  }
  static get listeners() {
    return [{
        "name": "keydown",
        "method": "handleKeyDown",
        "target": undefined,
        "capture": false,
        "passive": false
      }];
  }
}
