// @salla.sa/twilight-components/dist/collection/components/salla-user-settings/salla-user-settings.js
/*!
 * Crafted with â¤ by Salla
 */
import { Host, h } from '@stencil/core';
import Bullhorn from '../../assets/svg/bullhorn.svg';
import UserOff from '../../assets/svg/user-off.svg';
export class SallaUserSettings {
  constructor() {
    this.deactivateAccount = salla.lang.get('pages.profile.deactivate_account');
    this.promotionalMsgs = salla.lang.get('pages.profile.promotional_messages');
    this.deactivateDesc = salla.lang.get('pages.profile.deactivate_account_description');
    this.promotionalMsgsDesc = salla.lang.get('pages.profile.promotional_messages_description');
    this.sorryForLeavingText = salla.lang.get('pages.profile.warning_for_deactivate');
    this.warningText = salla.lang.get('pages.profile.sorry_for_leaving');
    this.keepAccount = salla.lang.get('pages.profile.keep_account');
    this.buttonLoading = false;
    this.isNotifiable = false;
    salla.lang.onLoaded(() => {
      this.deactivateAccount = salla.lang.get('pages.profile.deactivate_account');
      this.promotionalMsgs = salla.lang.get('pages.profile.promotional_messages');
      this.deactivateDesc = salla.lang.get('pages.profile.deactivate_account_description');
      this.promotionalMsgsDesc = salla.lang.get('pages.profile.promotional_messages_description');
      this.warningText = salla.lang.get('pages.profile.warning_for_deactivate');
      this.sorryForLeavingText = salla.lang.get('pages.profile.sorry_for_leaving');
      this.keepAccount = salla.lang.get('pages.profile.keep_account');
    });
  }
  toggleNotification(event) {
    salla.profile.updateSettings({ is_notifiable: event.target.checked });
  }
  capitalizeText(str) {
    return str.charAt(0).toUpperCase() + str.slice(1);
  }
  openDeactivateModal() {
    this.confirmationModal.setTitle(this.capitalizeText(this.deactivateAccount));
    this.confirmationModal.open();
  }
  async deleteAccount() {
    await salla.profile.delete().finally(() => this.confirmationModal.close());
  }
  render() {
    return (h(Host, { class: "s-user-settings-wrapper" }, h("div", { class: "s-user-settings-section" }, h("salla-list-tile", null, h("div", { slot: "icon", class: "s-user-settings-section-icon" }, h("span", { innerHTML: Bullhorn })), h("div", { slot: "title", class: "s-user-settings-section-title" }, this.capitalizeText(this.promotionalMsgs)), h("div", { slot: "subtitle", class: "s-user-settings-section-subtitle" }, this.capitalizeText(this.promotionalMsgsDesc)), h("div", { slot: 'action', class: "s-user-settings-section-action" }, h("label", { class: "s-toggle" }, h("input", { class: "s-toggle-input", checked: this.isNotifiable, onChange: (e) => this.toggleNotification(e), type: "checkbox" }), h("div", { class: "s-toggle-switcher" }))))), h("div", { class: "s-user-settings-section s-user-settings-section-deactivate-user" }, h("salla-list-tile", null, h("div", { slot: "icon", class: "s-user-settings-section-icon" }, h("span", { innerHTML: UserOff })), h("div", { slot: "title", class: "s-user-settings-section-title" }, this.capitalizeText(this.deactivateAccount)), h("div", { slot: "subtitle", class: "s-user-settings-section-subtitle" }, this.capitalizeText(this.deactivateDesc)), h("div", { slot: 'action', class: "s-user-settings-section-action" }, h("salla-button", { fill: "outline", color: "danger", shape: "btn", size: "medium", width: "normal", onClick: () => this.openDeactivateModal() }, this.capitalizeText(this.deactivateAccount))))), h("salla-modal", { width: "sm", subTitle: this.capitalizeText(this.sorryForLeavingText), ref: modal => this.confirmationModal = modal }, h("span", { slot: 'icon', class: "s-user-settings-confirmation-icon", innerHTML: UserOff }), h("div", { class: "s-user-settings-confirmation" }, h("div", { class: "s-user-settings-confirmation-warning" }, this.capitalizeText(this.warningText)), h("div", { class: "s-user-settings-confirmation-actions" }, h("salla-button", { width: "wide", onClick: () => this.confirmationModal.close() }, this.capitalizeText(this.keepAccount)), h("salla-button", { fill: 'outline', loading: this.buttonLoading, width: "wide", onClick: () => this.deleteAccount() }, this.capitalizeText(this.deactivateAccount)))))));
  }
  static get is() { return "salla-user-settings"; }
  static get originalStyleUrls() {
    return {
      "$": ["salla-user-settings.scss"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["salla-user-settings.css"]
    };
  }
  static get properties() {
    return {
      "isNotifiable": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Value used for handling notification toggle check box."
        },
        "attribute": "is-notifiable",
        "reflect": true,
        "defaultValue": "false"
      }
    };
  }
  static get states() {
    return {
      "deactivateAccount": {},
      "promotionalMsgs": {},
      "deactivateDesc": {},
      "promotionalMsgsDesc": {},
      "sorryForLeavingText": {},
      "warningText": {},
      "keepAccount": {},
      "buttonLoading": {}
    };
  }
}
