// @salla.sa/twilight-components/dist/collection/components/salla-skeleton/salla-skeleton.js
/*!
 * Crafted with ❤ by Salla
 */
import { h, Host } from '@stencil/core';
export class SallaSkeleton {
  constructor() {
    this.type = 'normal';
    this.width = '100%';
    this.height = '100%';
  }
  render() {
    const classes = {
      's-skeleton-item': true,
      's-skeleton-item-circular': this.type == 'circle',
    };
    return (h(Host, { class: "s-skeleton-wrapper", style: { width: this.width, height: this.height } }, h("div", { class: classes }, "\u00A0")));
  }
  static get is() { return "salla-skeleton"; }
  static get originalStyleUrls() {
    return {
      "$": ["salla-skeleton.scss"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["salla-skeleton.css"]
    };
  }
  static get properties() {
    return {
      "type": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "'circle' | 'normal'",
          "resolved": "\"circle\" | \"normal\"",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Set the shape type of the skeleton is it circle or normal"
        },
        "attribute": "type",
        "reflect": false,
        "defaultValue": "'normal'"
      },
      "width": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Set the skeleton width"
        },
        "attribute": "width",
        "reflect": false,
        "defaultValue": "'100%'"
      },
      "height": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Set the skeleton height"
        },
        "attribute": "height",
        "reflect": false,
        "defaultValue": "'100%'"
      }
    };
  }
}
