// @salla.sa/twilight-components/dist/esm/salla-products-slider.entry.js
/*!
 * Crafted with â¤ by Salla
 */
import { r as registerInstance, h, H as Host, g as getElement } from './index-957caa0e.js';

const sallaProductsSliderCss = "";

const SallaProductsSlider = class {
  constructor(hostRef) {
    registerInstance(this, hostRef);
    this.blockTitle = undefined;
    this.subTitle = undefined;
    this.sliderId = undefined;
    this.displayAllUrl = undefined;
    this.source = undefined;
    this.sourceValue = undefined;
    this.limit = undefined;
    this.autoplay = undefined;
    this.productsData = undefined;
    this.isReady = undefined;
    this.sourceValueIsValid = undefined;
    this.hasCustomComponent = undefined;
    this.apiUrl = '';
    this.parsedSourceValue = undefined;
    salla.onReady(() => {
      this.sourceValueIsValid = !!(this.getSource() || this.isSourceWithoutValue());
      if (!this.sourceValueIsValid) {
        salla.logger.warn(`source-value prop is required for source [${this.getSource()}]`);
        return;
      }
      this.hasCustomComponent = !!customElements.get('custom-salla-product-card');
    });
  }
  isSourceWithoutValue() {
    return ['offers', 'latest', 'sales'].includes(this.getSource());
  }
  getItemHTML(product) {
    if (this.hasCustomComponent) {
      return h("div", { class: "s-products-slider-card" }, h("custom-salla-product-card", { product: product, source: this.getSource(), "source-value": this.getSourceValue() }));
    }
    return h("div", { class: "s-products-slider-card" }, h("salla-product-card", { "show-quantity": this.getSource() == 'landing-page', "hide-add-btn": this.getSource() == 'landing-page', "shadow-on-hover": true, product: product }));
  }
  getSource() {
    return {
      'brands.single': 'brands',
      'product.index': 'categories',
      'product.index.latest': 'latest',
      'product.index.offers': 'offers',
      'product.index.search': 'search',
      'landing-page': 'landing-page',
      'product.index.tag': 'tags',
    }[this.source || salla.config.get('page.slug')] || this.source || 'latest';
  }
  getSourceValue() {
    let source = this.getSource();
    //todo:: enhance this code, it call JSON.parse too many times.
    if (!['search', 'json', 'offers', 'latest', 'sales', 'related'].includes(source)) {
      if (this.sourceValue && Array.isArray(JSON.parse(this.sourceValue)) && JSON.parse(this.sourceValue).length) {
        return JSON.parse(this.sourceValue);
      }
      if (this.sourceValue && typeof JSON.parse(this.sourceValue) == 'number') {
        return [JSON.parse(this.sourceValue)];
      }
      if (!this.sourceValue && ['categories', 'tags', 'brands'].includes(this.getSource())) {
        return [salla.config.get('page.id')];
      }
    }
    if (this.sourceValue) {
      return this.sourceValue;
    }
    if (source === 'search') {
      return (new URLSearchParams(window.location.search)).get('q');
    }
    return salla.config.get('page.id');
  }
  canRender() {
    return this.sourceValueIsValid && this.isReady;
  }
  componentWillLoad() {
    if (this.source === 'json') {
      this.productsData = this.getSourceValue();
      this.isReady = true;
      return;
    }
    return salla.product.fetch({
      source: this.getSource(),
      source_value: this.getSourceValue(),
      limit: this.limit,
    }).then(res => {
      this.productsData = res.data;
      this.isReady = true;
    });
  }
  render() {
    var _a;
    if ((this.getSource() == 'related' && !salla.config.get('store.settings.product.related_products_enabled')) || !this.canRender()) {
      return;
    }
    return (h(Host, { class: "s-products-slider-wrapper" }, h("salla-slider", { "auto-play": this.autoplay, class: "s-products-slider-slider", id: this.sliderId || `s-products-slider-${Math.random().toString(36).substr(2, 9)}`, type: "carousel", "block-title": this.blockTitle, "block-subTitle": this.subTitle, "display-all-url": this.displayAllUrl }, h("div", { slot: "items" }, (_a = this.productsData) === null || _a === void 0 ? void 0 : _a.map(product => this.getItemHTML(product))))));
  }
  get host() { return getElement(this); }
};
SallaProductsSlider.style = sallaProductsSliderCss;

export { SallaProductsSlider as salla_products_slider };
