// @salla.sa/twilight-components/dist/esm/salla-comment-form.entry.js
/*!
 * Crafted with â¤ by Salla
 */
import { r as registerInstance, h, H as Host } from './index-957caa0e.js';

const sallaCommentFormCss = ":host{display:block}";

const SallaCommentForm = class {
  constructor(hostRef) {
    registerInstance(this, hostRef);
    this.type = undefined;
    this.showAvatar = undefined;
    this.itemId = undefined;
    this.placeholder = salla.lang.get('blocks.comments.placeholder');
    this.submitText = salla.lang.get('blocks.comments.submit');
    this.canComment = undefined;
    salla.lang.onLoaded(() => {
      this.placeholder = salla.lang.get('blocks.comments.placeholder');
      this.submitText = salla.lang.get('blocks.comments.submit');
    });
    salla.onReady(() => {
      this.canComment = salla.config.get('user.can_comment');
      this.itemId = salla.config.get('page.id');
      this.type = salla.url.is_page('page-single') ? 'page' : 'product';
    });
  }
  submit() {
    if (!this.commentForm.reportValidity()) {
      salla.log('CommentForm:: validation error!');
      return;
    }
    this.submitBtn.load()
      .then(() => salla.comment.add({ id: this.itemId, comment: this.commentField.value, type: this.type }))
      .finally(() => this.submitBtn.stop);
  }
  render() {
    return (h(Host, null, !!this.canComment ? h("form", { ref: frm => this.commentForm = frm }, h("div", { class: "s-comment-form-wrapper" }, this.showAvatar ?
      h("img", { class: "s-comment-form-avatar", src: salla.config.get('user.avatar'), alt: "user avatar" }) : '', h("div", { class: "s-comment-form-content" }, h("textarea", { cols: 30, rows: 5, minlength: "4", maxlength: "500", ref: field => this.commentField = field, placeholder: this.placeholder, class: "s-comment-form-input", required: true }), h("br", null), h("div", { class: "s-comment-form-action" }, h("salla-button", { ref: btn => this.submitBtn = btn, "loader-position": 'center', onClick: () => this.submit() }, this.submitText))))) : ''));
  }
};
SallaCommentForm.style = sallaCommentFormCss;

export { SallaCommentForm as salla_comment_form };
