// @salla.sa/twilight-components/dist/cjs/salla-product-options.cjs.entry.js
/*!
 * Crafted with ❤ by Salla
 */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-e85bd084.js');
const camera = require('./camera-058ddeeb.js');

var DisplayType;
(function (DisplayType) {
  DisplayType["COLOR"] = "color";
  DisplayType["DATE"] = "date";
  DisplayType["DATETIME"] = "datetime";
  DisplayType["DONATION"] = "donation";
  DisplayType["IMAGE"] = "image";
  DisplayType["MULTIPLE_OPTIONS"] = "multiple-options";
  DisplayType["NUMBER"] = "number";
  DisplayType["SINGLE_OPTION"] = "single-option";
  DisplayType["SPLITTER"] = "splitter";
  DisplayType["TEXT"] = "text";
  DisplayType["TEXTAREA"] = "textarea";
  DisplayType["THUMBNAIL"] = "thumbnail";
  DisplayType["TIME"] = "time";
  DisplayType["RADIO"] = "radio";
  DisplayType["CHECKBOX"] = "checkbox";
  DisplayType["MAP"] = "map";
  DisplayType["FILE"] = "file";
  DisplayType["COLOR_PICKER"] = "color_picker";
})(DisplayType || (DisplayType = {}));
var Currency;
(function (Currency) {
  Currency["Sar"] = "SAR";
})(Currency || (Currency = {}));

const CheckCircleIcon = `<!-- Generated by IcoMoon.io -->
<svg version="1.1" xmlns="http://www.w3.org/2000/svg" width="32" height="32" viewBox="0 0 32 32">
<title>check</title>
<path d="M27.521 6.976c-0.569-0.472-1.407-0.393-1.879 0.171l-12.567 15.080-7.003-4.668c-0.615-0.411-1.441-0.244-1.849 0.369-0.409 0.612-0.244 1.441 0.369 1.849l8 5.333c0.227 0.149 0.484 0.223 0.739 0.223 0.384 0 0.763-0.165 1.027-0.48l13.333-16c0.471-0.565 0.393-1.407-0.171-1.877z"></path>
</svg>
`;

const FileIcon = `<!-- Generated by IcoMoon.io -->
<svg version="1.1" xmlns="http://www.w3.org/2000/svg" width="32" height="32" viewBox="0 0 32 32">
<title>file-upload</title>
<path d="M21.333 24c0.341 0 0.683-0.131 0.943-0.391 0.521-0.521 0.521-1.364 0-1.885l-5.333-5.333c-0.123-0.123-0.271-0.22-0.433-0.288-0.327-0.135-0.693-0.135-1.019 0-0.163 0.068-0.311 0.165-0.433 0.288l-5.333 5.333c-0.521 0.521-0.521 1.364 0 1.885s1.364 0.521 1.885 0l3.057-3.057v10.115c0 0.736 0.597 1.333 1.333 1.333s1.333-0.597 1.333-1.333v-10.115l3.057 3.057c0.26 0.26 0.601 0.391 0.943 0.391zM28.943 9.724l-9.333-9.333c-0.249-0.251-0.589-0.391-0.943-0.391h-12c-2.205 0-4 1.795-4 4v24c0 2.205 1.795 4 4 4h4c0.736 0 1.333-0.597 1.333-1.333s-0.597-1.333-1.333-1.333h-4c-0.735 0-1.333-0.599-1.333-1.333v-24c0-0.735 0.599-1.333 1.333-1.333h11.448l8.552 8.552v16.781c0 0.735-0.599 1.333-1.333 1.333h-4c-0.736 0-1.333 0.597-1.333 1.333s0.597 1.333 1.333 1.333h4c2.205 0 4-1.795 4-4v-17.333c0-0.353-0.14-0.693-0.391-0.943z"></path>
</svg>
`;

const sallaProductOptionsCss = "";

const SallaProductOptions = class {
  constructor(hostRef) {
    index.registerInstance(this, hostRef);
    this.changed = index.createEvent(this, "changed", 7);
    this.fileTypes = {
      pdf: 'application/pdf',
      png: 'image/png',
      jpg: 'image/jpeg',
      word: 'application/doc,application/ms-doc,application/msword,application/vnd.openxmlformats-officedocument.wordprocessingml.document',
      exl: 'application/excel,application/vnd.ms-excel,application/x-excel,application/x-msexcel,application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
      txt: 'text/plain',
    };
    this.outSkus = [];
    this.optionsData = undefined;
    this.outOfStockText = '';
    this.donationAmount = salla.lang.get('pages.products.donation_amount');
    this.selectedOptions = [];
    this.canDisabled = undefined;
    this.selectedSkus = undefined;
    this.selectedOutSkus = undefined;
    this.productId = salla.config.get('page.id');
    this.options = undefined;
    this.canDisabled = !salla.config.get('store.settings.products.notify_options_availability');
    salla.lang.onLoaded(() => {
      this.outOfStockText = salla.lang.get("pages.products.out_of_stock");
      this.donationAmount = salla.lang.get('pages.products.donation_amount');
    });
    if (this.options) {
      try {
        this.setOptionsData(JSON.parse(this.options));
        return;
      }
      catch (e) {
        salla.log('Bad json passed via options prop');
      }
    }
    if (!Array.isArray(this.optionsData)) {
      salla.log('Options is not an array[] ---> ', this.optionsData);
      this.setOptionsData([]);
    }
    if (this.productId && !salla.url.is_page('cart')) {
      salla.api.product.getDetails(this.productId, ['options']).then(resp => this.setOptionsData(resp.data.options));
    }
  }
  setOptionsData(optionsData) {
    var _a, _b;
    this.optionsData = optionsData;
    let that = this;
    (_b = (_a = this.optionsData[0]) === null || _a === void 0 ? void 0 : _a.details) === null || _b === void 0 ? void 0 : _b.forEach(function (detail) {
      Object.entries(detail.skus_availability || {})
        .filter(sku => !sku[1])
        .map(sku => that.outSkus.push(Number(sku[0])));
    });
  }
  /**
   * Get the id's of the selected options.
   * */
  async getSelectedOptionsData() {
    let selectedOptions = {};
    let formData = this.host.getElementSallaData();
    formData.forEach(function (value, key) {
      key.startsWith('options[') && (selectedOptions[key.replace('options[', '').replace(']', '')] = value);
    });
    return selectedOptions;
  }
  /**
   * Report options form validity.
   * */
  async reportValidity() {
    let requiredElements = this.host.querySelectorAll('[required]');
    let pass = true;
    for (let i = 0; i < requiredElements.length; i++) {
      //if there is only one invalid option, return false
      if ('reportValidity' in requiredElements[i] && !requiredElements[i].reportValidity()) {
        pass = false;
      }
    }
    return pass;
  }
  /**
   * Return true if there is any out of stock options are selected and vise versa.
   * */
  async hasOutOfStockOption() {
    return this.selectedOptions.some(option => option.is_out) || (this.selectedSkus.length && this.selectedSkus.every(sku => this.outSkus.includes(sku)));
  }
  /**
   * Get selected options.
   * */
  async getSelectedOptions() {
    return this.selectedOptions;
  }
  /**
   * Get a specific option by its id.
   * */
  async getOption(option_id) {
    return this.optionsData.find(option => option.id === option_id);
  }
  // @ts-ignore
  invalidHandler(event, option) {
    const closestProductOption = event.target.closest('.s-product-options-option');
    if (!salla.url.is_page('cart')) {
      closestProductOption.scrollIntoView({ behavior: 'smooth', block: 'center' });
    }
    closestProductOption.classList.add('s-product-options-option-error');
  }
  changedHandler(event, option) {
    let data = { event: event, option: option, detail: null };
    if (option.details) {
      let detail = option.details.find((detail) => {
        return Number(detail.id) === Number(event.target.value);
      });
      data.detail = detail;
    }
    let optionElement = event.target.closest('.s-product-options-option');
    if (event.target.value
      || ((option.type == DisplayType.FILE || option.type == DisplayType.IMAGE) && event.type === 'added')
      || (option.type == DisplayType.MAP && event.type === 'selected' && (event.target.lat && event.target.lng))) {
      setTimeout(() => {
        optionElement.classList.remove('s-product-options-option-error');
      }, 200);
    }
    const index = this.selectedOptions.findIndex(option => option.option_id === data.option.id);
    index > -1 ? this.selectedOptions[index] = Object.assign(Object.assign({}, data.detail), { option_id: data.option.id }) : this.selectedOptions.push(Object.assign(Object.assign({}, data.detail), { option_id: data.option.id }));
    this.setSelectedSkus();
    this.handleRequiredMultipleOptions(option);
    this.changed.emit(data);
    salla.event.emit('product-options::change', data);
  }
  /**
   * loop throw all selected details, then get common sku, if it's only one, means we selected all of them;
   */
  setSelectedSkus() {
    this.selectedSkus = this.selectedOptions.map(detail => Object.keys(detail.skus_availability || {}))
      .reduce((p, c) => p.filter(e => c.includes(e)))
      .map(sku => Number(sku));
  }
  handleRequiredMultipleOptions(option) {
    if (option.type !== DisplayType.MULTIPLE_OPTIONS || !option.required) {
      return;
    }
    const optionContainer = this.host.querySelector(`[data-option-id="${option.id}"]`);
    const hasChecked = optionContainer.querySelectorAll('input:checked').length;
    optionContainer.querySelectorAll('input').forEach(input => input.toggleAttribute('required', !hasChecked));
  }
  getLatLng(value, type) {
    return value
      ? value.split(',')[type == 'lat' ? 0 : 1]
      : '';
  }
  getDisplayForType(option) {
    if (this[`${option.type}Option`]) {
      return this[`${option.type}Option`](option);
    }
    if (option.type === DisplayType.COLOR_PICKER) {
      return this.colorPickerOption(option);
    }
    if (option.type === DisplayType.MULTIPLE_OPTIONS) {
      return this.multipleOptions(option);
    }
    if (option.type === DisplayType.SINGLE_OPTION) {
      return this.singleOption(option);
    }
    salla.log(`Couldn't find options type(${option.type})😢`);
    return '';
  }
  getOptionShownWhen(option) {
    return option.visibility_condition
      ? { "data-show-when": `options[${option.visibility_condition.option}] ${option.visibility_condition.operator} ${option.visibility_condition.value}` }
      : {};
  }
  //we need the cart Id for productOption Image
  componentWillLoad() {
    this.outOfStockText = salla.lang.get('pages.products.out_of_stock');
    return salla.api.cart.getCurrentCartId();
  }
  render() {
    var _a;
    if (((_a = this.optionsData) === null || _a === void 0 ? void 0 : _a.length) == 0) {
      return;
    }
    return (index.h(index.Host, { class: "s-product-options-wrapper" }, index.h("salla-conditional-fields", null, this.optionsData.map((option) => index.h("div", Object.assign({ class: `s-product-options-option-container${option.visibility_condition ? ' hidden' : ''}`, "data-option-id": option.id }, this.getOptionShownWhen(option)), option.name == 'splitter' ?
      this.splitterOption()
      : index.h("div", { class: "s-product-options-option", "data-option-type": option.type, "data-option-required": `${option.required}` }, index.h("label", { htmlFor: 'options[' + option.id + ']', class: "s-product-options-option-label" }, index.h("b", null, option.name, option.required && index.h("span", null, " * "), " "), index.h("small", null, option.placeholder)), index.h("div", { class: "s-product-options-option-content" }, this.getDisplayForType(option))))))));
  }
  //@ts-ignore
  donationOption(option, product) {
    return index.h("div", { class: "s-product-options-donation-wrapper" }, option.donation ?
      index.h("div", { class: "s-product-options-donation-progress" }, index.h("salla-progress-bar", { donation: option.donation }))
      : '', index.h("div", { class: "s-product-options-donation-input-group" }, index.h("input", { type: "text", id: "donating-amount", name: "donating_amount", class: "s-form-control", value: option.value, required: true, placeholder: option.placeholder, onInput: e => salla.helpers.inputDigitsOnly(e.target), onBlur: e => this.changedHandler(e, option), onInvalid: (e) => this.invalidHandler(e, option) }), index.h("span", { class: "s-product-options-donation-amount-currency" }, salla.config.currency(salla.config.get('user.currency_code')).symbol)));
  }
  fileUploader(option, additions = null) {
    return index.h("salla-file-upload", Object.assign({}, (additions || {}), { "payload-name": "file", value: option.value, "instant-upload": true, name: `options[${option.id}]`, required: option.required, height: "120px", onAdded: (e) => this.changedHandler(e, option), url: salla.cart.api.getUploadImageEndpoint(), "form-data": { cart_item_id: this.productId, product_id: this.productId }, onInvalidInput: (e) => this.invalidHandler(e, option), class: { "s-product-options-image-input": true, required: option.required } }), index.h("div", { class: "s-product-options-filepond-placeholder" }, index.h("span", { class: "s-product-options-filepond-placeholder-icon", innerHTML: additions.accept && additions.accept.split(',').every(type => type.includes('image'))
        ? camera.CameraIcon
        : FileIcon }), index.h("p", { class: "s-product-options-filepond-placeholder-text" }, salla.lang.get('common.uploader.drag_and_drop')), index.h("span", { class: "filepond--label-action" }, salla.lang.get('common.uploader.browse'))));
  }
  //@ts-ignore
  imageOption(option) {
    return this.fileUploader(option, { accept: 'image/png,image/jpeg,image/jpg,image/gif' });
  }
  //@ts-ignore
  fileOption(option) {
    let types = option.details.map(detail => this.fileTypes[detail.name]).filter(Boolean);
    return (types === null || types === void 0 ? void 0 : types.length)
      ? this.fileUploader(option, { accept: types.join(',') })
      : 'File types not selected.';
  }
  //@ts-ignore
  numberOption(option) {
    return index.h("input", { type: "text", value: option.value, class: "s-form-control", required: option.required, name: `options[${option.id}]`, placeholder: option.placeholder, onBlur: e => this.changedHandler(e, option), onInvalid: (e) => this.invalidHandler(e, option), onInput: e => salla.helpers.inputDigitsOnly(e.target) });
  }
  //@ts-ignore
  splitterOption() {
    return index.h("div", { class: "s-product-options-splitter" });
  }
  //@ts-ignore
  textOption(option) {
    return index.h("div", { class: "s-product-options-text" }, index.h("input", { type: "text", value: option.value, class: 's-form-control', required: option.required, name: `options[${option.id}]`, placeholder: option.placeholder, onInvalid: (e) => this.invalidHandler(e, option), onInput: e => this.changedHandler(e, option) }));
  }
  //@ts-ignore
  textareaOption(option) {
    //todo::remove mt-1 class, and if it's okay to remove the tag itself will be great
    return index.h("div", { class: "s-product-options-textarea" }, index.h("div", { class: "mt-1" }, index.h("textarea", { rows: 4, value: option.value, class: "s-form-control", required: option.required, id: `options[${option.id}]`, name: `options[${option.id}]`, placeholder: option.placeholder, onInvalid: (e) => this.invalidHandler(e, option), onInput: (e) => this.changedHandler(e, option) })));
  }
  //@ts-ignore
  mapOption(option) {
    return index.h("salla-map", { zoom: 15, lat: this.getLatLng(option.value, 'lat'), lng: this.getLatLng(option.value, 'lng'), name: `options[${option.id}]`, searchable: true, required: option.required, onInvalidInput: (e) => this.invalidHandler(e, option), onSelected: e => this.changedHandler(e, option) });
  }
  colorPickerOption(option) {
    return index.h("salla-color-picker", { onSubmitted: e => this.changedHandler(e, option), name: `options[${option.id}]`, required: option.required, onInvalidInput: (e) => this.invalidHandler(e, option), color: option.value });
  }
  /**
   * ============= Date Time options =============
   */
  //@ts-ignore
  timeOption(option) {
    return index.h("salla-datetime-picker", { noCalendar: true, enableTime: true, dateFormat: "h:i K", value: option.value, placeholder: option.name, required: option.required, name: `options[${option.id}]`, class: "s-product-options-time-element", onInvalidInput: (e) => this.invalidHandler(e, option), onPicked: e => this.changedHandler(e, option) });
  }
  //@ts-ignore
  dateOption(option) {
    //todo:: consider date-range @see https://github.com/SallaApp/theme-raed/blob/master/src/assets/js/partials/product-options.js#L8-L23
    return index.h("div", { class: "s-product-options-date-element" }, index.h("salla-datetime-picker", { value: option.value, placeholder: option.name, required: option.required, minDate: new Date(), name: `options[${option.id}]`, onInvalidInput: (e) => this.invalidHandler(e, option), onPicked: e => this.changedHandler(e, option) }));
  }
  //@ts-ignore
  datetimeOption(option) {
    //todo:: consider date-range @see https://github.com/SallaApp/theme-raed/blob/master/src/assets/js/partials/product-options.js#L8-L23
    return index.h("div", { class: "s-product-options-datetime-element" }, index.h("salla-datetime-picker", { enableTime: true, value: option.value, dateFormat: "Y-m-d G:i:K", placeholder: option.name, required: option.required, name: `options[${option.id}]`, maxDate: option.to_date_time, minDate: option.from_date_time, onInvalidInput: (e) => this.invalidHandler(e, option), onPicked: e => this.changedHandler(e, option) }));
  }
  /**
   * ============= Advanced options =============
   */
  getOptionDetailName(detail, outOfStock = true, optionType) {
    if (optionType && optionType == DisplayType.COLOR) {
      return detail.name
        + ((outOfStock && this.isOptionDetailOut(detail)) ? ` <br/> <p> ${this.outOfStockText} </p>` : '')
        + (detail.additional_price ? ` <p> (${salla.money(detail.additional_price)}) </p>` : '');
    }
    return detail.name
      + ((outOfStock && this.isOptionDetailOut(detail)) ? ` - ${this.outOfStockText}` : '')
      + (detail.additional_price ? ` (${salla.money(detail.additional_price)})` : '');
  }
  isOptionDetailOut(detail) {
    var _a;
    if (detail.is_out || !detail.skus_availability || !((_a = this.selectedSkus) === null || _a === void 0 ? void 0 : _a.length)) {
      return detail.is_out;
    }
    let isDetailSelected = this.selectedOptions.filter(option => option.id == detail.id).length;
    //if the current options is the only selected option, so we are sure that it's not out, because there is no other options selected yet
    if (isDetailSelected && this.selectedOptions.length == 1) {
      return false;
    }
    //if current details has sku in the possible outSkus it's out for sure
    if (isDetailSelected) {
      //here we will get the possible outSkus for current selected options
      let outSelectableSkus = this.selectedSkus.filter(sku => this.outSkus.includes(sku));
      return Object.keys(detail.skus_availability).some(sku => outSelectableSkus.includes(Number(sku)));
    }
    return this.selectedOptions.some(option => option.is_out && option.option_id !== detail.option_id);
  }
  singleOption(option) {
    return index.h("div", null, index.h("select", { name: `options[${option.id}]`, required: option.required, class: "s-form-control", onInvalid: (e) => this.invalidHandler(e, option), onChange: e => this.changedHandler(e, option) }, index.h("option", { value: "" }, option.placeholder), option === null || option === void 0 ? void 0 :
      option.details.map((detail) => {
        return index.h("option", { value: detail.id, disabled: this.canDisabled && this.isOptionDetailOut(detail), selected: detail.is_selected }, this.getOptionDetailName(detail));
      })));
  }
  multipleOptions(option) {
    return index.h("div", { class: { "s-product-options-multiple-options-wrapper": true, 'required': option.required } }, option === null || option === void 0 ? void 0 : option.details.map((detail) => {
      return index.h("div", null, index.h("input", { type: "checkbox", value: detail.id, disabled: this.isOptionDetailOut(detail), checked: detail.is_selected, required: option.required, name: `options[${option.id}][]`, id: `field-${option.id}-${detail.id}`, onChange: (e) => this.changedHandler(e, option), onInvalid: (e) => this.invalidHandler(e, option), "aria-describedby": `options[${option.id}]-description` }), index.h("label", { htmlFor: `field-${option.id}-${detail.id}` }, this.getOptionDetailName(detail)));
    }));
  }
  //@ts-ignore
  colorOption(option) {
    return index.h("fieldset", { class: "s-product-options-colors-wrapper" }, option === null || option === void 0 ? void 0 : option.details.map((detail) => index.h("div", { class: "s-product-options-colors-item" }, index.h("input", { type: "radio", value: detail.id, required: option.required, checked: detail.is_selected, name: `options[${option.id}]`, disabled: this.canDisabled && this.isOptionDetailOut(detail), id: `color-${this.productId}-${option.id}-${detail.id}`, onInvalid: (e) => this.invalidHandler(e, option), onChange: e => this.changedHandler(e, option) }), index.h("label", { htmlFor: `color-${this.productId}-${option.id}-${detail.id}` }, index.h("span", { style: { "background-color": detail.color } }), index.h("div", { innerHTML: this.getOptionDetailName(detail, true, option.type) })))));
  }
  //@ts-ignore
  thumbnailOption(option) {
    return index.h("div", { class: "s-product-options-thumbnails-wrapper" }, option.details.map((detail) => {
      return index.h("div", null, index.h("input", { type: "radio", value: detail.id, "data-itemid": detail.id, required: option.required, checked: detail.is_selected, name: `options[${option.id}]`, "data-img-id": detail.option_value, disabled: this.canDisabled && this.isOptionDetailOut(detail), id: `option_${this.productId}-${option.id}_${detail.id}`, onInvalid: (e) => this.invalidHandler(e, option), onChange: (e) => this.changedHandler(e, option) }), index.h("label", { htmlFor: `option_${this.productId}-${option.id}_${detail.id}`, "data-img-id": detail.option_value, class: "go-to-slide" }, index.h("img", { "data-src": detail.image, src: detail.image, title: detail.name, alt: detail.name }), index.h("span", { innerHTML: CheckCircleIcon, class: "s-product-options-thumbnails-icon" }), this.isOptionDetailOut(detail) ?
        [
          index.h("small", { class: "s-product-options-thumbnails-stock-badge" }, this.outOfStockText),
          this.canDisabled ? index.h("div", { class: "s-product-options-thumbnails-badge-overlay" }) : '',
        ]
        : ''), index.h("p", null, this.getOptionDetailName(detail, false), " "));
    }));
  }
  get host() { return index.getElement(this); }
};
SallaProductOptions.style = sallaProductOptionsCss;

exports.salla_product_options = SallaProductOptions;
