// @salla.sa/twilight-components/dist/cjs/salla-comment-form.cjs.entry.js
/*!
 * Crafted with â¤ by Salla
 */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-e85bd084.js');

const sallaCommentFormCss = ":host{display:block}";

const SallaCommentForm = class {
  constructor(hostRef) {
    index.registerInstance(this, hostRef);
    this.type = undefined;
    this.showAvatar = undefined;
    this.itemId = undefined;
    this.placeholder = salla.lang.get('blocks.comments.placeholder');
    this.submitText = salla.lang.get('blocks.comments.submit');
    this.canComment = undefined;
    salla.lang.onLoaded(() => {
      this.placeholder = salla.lang.get('blocks.comments.placeholder');
      this.submitText = salla.lang.get('blocks.comments.submit');
    });
    salla.onReady(() => {
      this.canComment = salla.config.get('user.can_comment');
      this.itemId = salla.config.get('page.id');
      this.type = salla.url.is_page('page-single') ? 'page' : 'product';
    });
  }
  submit() {
    if (!this.commentForm.reportValidity()) {
      salla.log('CommentForm:: validation error!');
      return;
    }
    this.submitBtn.load()
      .then(() => salla.comment.add({ id: this.itemId, comment: this.commentField.value, type: this.type }))
      .finally(() => this.submitBtn.stop);
  }
  render() {
    return (index.h(index.Host, null, !!this.canComment ? index.h("form", { ref: frm => this.commentForm = frm }, index.h("div", { class: "s-comment-form-wrapper" }, this.showAvatar ?
      index.h("img", { class: "s-comment-form-avatar", src: salla.config.get('user.avatar'), alt: "user avatar" }) : '', index.h("div", { class: "s-comment-form-content" }, index.h("textarea", { cols: 30, rows: 5, minlength: "4", maxlength: "500", ref: field => this.commentField = field, placeholder: this.placeholder, class: "s-comment-form-input", required: true }), index.h("br", null), index.h("div", { class: "s-comment-form-action" }, index.h("salla-button", { ref: btn => this.submitBtn = btn, "loader-position": 'center', onClick: () => this.submit() }, this.submitText))))) : ''));
  }
};
SallaCommentForm.style = sallaCommentFormCss;

exports.salla_comment_form = SallaCommentForm;
