// @salla.sa/twilight-components/dist/collection/components/salla-products-slider/salla-products-slider.js
/*!
 * Crafted with â¤ by Salla
 */
import { Host, h } from '@stencil/core';
//todo:: extends this component from salla-products-list or the opposite
export class SallaProductsSlider {
  constructor() {
    this.blockTitle = undefined;
    this.subTitle = undefined;
    this.sliderId = undefined;
    this.displayAllUrl = undefined;
    this.source = undefined;
    this.sourceValue = undefined;
    this.limit = undefined;
    this.productsData = undefined;
    this.isReady = undefined;
    this.sourceValueIsValid = undefined;
    this.hasCustomComponent = undefined;
    this.apiUrl = '';
    this.parsedSourceValue = undefined;
    salla.onReady(() => {
      this.sourceValueIsValid = !!(this.getSource() || this.isSourceWithoutValue());
      if (!this.sourceValueIsValid) {
        salla.logger.warn(`source-value prop is required for source [${this.getSource()}]`);
        return;
      }
      this.hasCustomComponent = !!customElements.get('custom-salla-product-card');
    });
  }
  isSourceWithoutValue() {
    return ['offers', 'latest', 'sales'].includes(this.getSource());
  }
  getItemHTML(product) {
    if (this.hasCustomComponent) {
      return h("div", { class: "s-products-slider-card" }, h("custom-salla-product-card", { product: product, source: this.getSource(), "source-value": this.getSourceValue() }));
    }
    return h("div", { class: "s-products-slider-card" }, h("salla-product-card", { "show-quantity": this.getSource() == 'landing-page', "hide-add-btn": this.getSource() == 'landing-page', "shadow-on-hover": true, product: product }));
  }
  getSource() {
    return {
      'brands.single': 'brands',
      'product.index': 'categories',
      'product.index.latest': 'latest',
      'product.index.offers': 'offers',
      'product.index.search': 'search',
      'landing-page': 'landing-page',
      'product.index.tag': 'tags',
    }[this.source || salla.config.get('page.slug')] || this.source || 'latest';
  }
  getSourceValue() {
    let source = this.getSource();
    //todo:: enhance this code, it call JSON.parse too many times.
    if (!['search', 'json', 'offers', 'latest', 'sales', 'related'].includes(source)) {
      if (this.sourceValue && Array.isArray(JSON.parse(this.sourceValue)) && JSON.parse(this.sourceValue).length) {
        return JSON.parse(this.sourceValue);
      }
      if (this.sourceValue && typeof JSON.parse(this.sourceValue) == 'number') {
        return [JSON.parse(this.sourceValue)];
      }
      if (!this.sourceValue && ['categories', 'tags', 'brands'].includes(this.getSource())) {
        return [salla.config.get('page.id')];
      }
    }
    if (this.sourceValue) {
      return this.sourceValue;
    }
    if (source === 'search') {
      return (new URLSearchParams(window.location.search)).get('q');
    }
    return salla.config.get('page.id');
  }
  canRender() {
    return this.sourceValueIsValid && this.isReady;
  }
  componentWillLoad() {
    if (this.source === 'json') {
      this.productsData = this.getSourceValue();
      this.isReady = true;
      return;
    }
    return salla.product.fetch({
      source: this.getSource(),
      source_value: this.getSourceValue(),
      limit: this.limit,
    }).then(res => {
      this.productsData = res.data;
      this.isReady = true;
    });
  }
  render() {
    var _a;
    if ((this.getSource() == 'related' && !salla.config.get('store.settings.product.related_products_enabled')) || !this.canRender()) {
      return;
    }
    return (h(Host, { class: "s-products-slider-wrapper" }, h("salla-slider", { class: "s-products-slider-slider", id: this.sliderId || `s-products-slider-${Math.random().toString(36).substr(2, 9)}`, type: "carousel", "block-title": this.blockTitle, "block-subTitle": this.subTitle, "display-all-url": this.displayAllUrl }, h("div", { slot: "items" }, (_a = this.productsData) === null || _a === void 0 ? void 0 : _a.map(product => this.getItemHTML(product))))));
  }
  static get is() { return "salla-products-slider"; }
  static get originalStyleUrls() {
    return {
      "$": ["salla-products-slider.scss"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["salla-products-slider.css"]
    };
  }
  static get properties() {
    return {
      "blockTitle": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "type",
              "text": "{string}"
            }, {
              "name": "default",
              "text": "''"
            }],
          "text": "Title of the block - works only if slider is true"
        },
        "attribute": "block-title",
        "reflect": false
      },
      "subTitle": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "type",
              "text": "{string}"
            }, {
              "name": "default",
              "text": "''"
            }],
          "text": "Sub title of the block - works only if slider is true"
        },
        "attribute": "sub-title",
        "reflect": false
      },
      "sliderId": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "type",
              "text": "{string}"
            }, {
              "name": "default",
              "text": "''"
            }],
          "text": "Slider Id, if not provided will be generated automatically"
        },
        "attribute": "slider-id",
        "reflect": false
      },
      "displayAllUrl": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "type",
              "text": "{string}"
            }, {
              "name": "default",
              "text": "''"
            }],
          "text": "Display 'ALL' URL"
        },
        "attribute": "display-all-url",
        "reflect": false
      },
      "source": {
        "type": "string",
        "mutable": true,
        "complexType": {
          "original": "'categories' | 'latest' | 'related' | 'brands' | 'json' | 'tags' | 'selected' | 'offers' | 'landing-page'",
          "resolved": "\"brands\" | \"categories\" | \"json\" | \"landing-page\" | \"latest\" | \"offers\" | \"related\" | \"selected\" | \"tags\"",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "type",
              "text": "{string}"
            }, {
              "name": "default",
              "text": "''"
            }],
          "text": "Source of the products, if api will get the products from the API, if json will get the products from the products prop"
        },
        "attribute": "source",
        "reflect": true
      },
      "sourceValue": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "type",
              "text": "{string}"
            }],
          "text": "The source value, cloud be different values as following:\n- array of ids when `source` in ['categories', 'brands', 'tags', 'selected']\n- products payload when `source` = 'json'\n- product_id when `source` = 'related'"
        },
        "attribute": "source-value",
        "reflect": false
      },
      "limit": {
        "type": "number",
        "mutable": true,
        "complexType": {
          "original": "number",
          "resolved": "number",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Limit for number of products in the list."
        },
        "attribute": "limit",
        "reflect": false
      }
    };
  }
  static get states() {
    return {
      "productsData": {},
      "isReady": {},
      "sourceValueIsValid": {},
      "hasCustomComponent": {},
      "apiUrl": {},
      "parsedSourceValue": {}
    };
  }
  static get elementRef() { return "host"; }
}
